README.md
ReadMe
========

Daniel Hopkins and Gary King

## Description

The ReadMe software package for R takes as input a set of text
documents (such as speeches, blog posts, newspaper articles, judicial
opinions, movie reviews, etc.), a categorization scheme chosen by the
user (e.g., ordered positive to negative sentiment ratings, unordered
policy topics, or any other mutually exclusive and exhaustive set of
categories), and a small subset of text documents hand classified into
the given categories. If used properly, ReadMe will report,
within sampling error of the truth, the proportion of documents within
each of the given categories among those not hand coded. ReadMe
computes the distribution within categories without the more error
prone intermediate step of classifying individual documents. Various
other procedures are included to make processing text easy.
See Daniel Hopkins and Gary King, "A Method of Automated Nonparametric Content
Analysis for Social Science," American Journal of Political
Science, <http://gking.harvard.edu/files/abs/words-abs.shtml>


## Example

```r
library(ReadMe)

oldwd <- getwd()
setwd(system.file("demofiles/clintonposts",
      package = "ReadMe"))

undergrad.results <- undergrad(sep = ',')

undergrad.preprocess <- preprocess(undergrad.results)

readme.results <- readme(undergrad.preprocess)
setwd(oldwd)
```

<a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/3.0/us/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-nd/3.0/us/80x15.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/3.0/us/">Creative Commons Attribution-NonCommercial-NoDerivs 3.0 United States License</a>.


NEWS.md
# ReadMe 0.99837

* Removed Design package dependency.

* Added a `NEWS.md` file to track changes to the package.





R/prototype.R
#' Remove non-variant
#' @param tab [to complete in]
#'
#' @export

remove.nonvariant <- function(tab)
{
    ncols <- ncol(tab)
    i <- 3
    while(i < ncols){
        i <- i + 1
        m <- mean(as.numeric(tab[,i]))
        if (m == 1 | m == 0) {
          tab <- tab[,-i]
          i <- i - 1
          ncols <- ncols - 1
        }
    }
    return (tab)
}

#' Remove make binary
#' @param tab [to complete in]
#'
#' @export

make.binary <- function(tab)
{
  r <- nrow(tab)
  c <- ncol(tab)
  for(i in 1:r)
  {
     print(r)
     for (j in 1:c)
     {
         if(is.na(tab[i,j]))
            tab[i,j] <- 0
         if(tab[i,j] != 0)
           tab[i,j] <- 1
     }
  }

  return (tab)
}

#' @importFrom VA va
#' @export

readme <- function(undergradlist = list(), trainingset = NULL, testset = NULL,
                   formula = NULL, features = NULL, n.subset =
                   NULL, prob.wt = NULL, boot.se = NULL, nboot = NULL,
                   printit = NULL)
{
  if(is.null(testset))
    testset <- undergradlist$testset

  if(is.null(trainingset))
    trainingset <- undergradlist$trainingset

  hospital <- trainingset[,-c(1,3)]
  community <- testset[,-c(1,3)]

  if(is.null(formula))
    formula <- undergradlist$formula

  if(is.null(features))
    features <- undergradlist$features

  if(is.null(n.subset))
    n.subset <- undergradlist$n.subset

  if(is.null(prob.wt))
    prob.wt <- undergradlist$prob.wt

  if(is.null(boot.se))
    boot.se <- undergradlist$boot.se

  if(is.null(printit))
    printit <- undergradlist$printit

  if(is.null(nboot))
    nboot <- undergradlist$nboot

  return (va(formula = formula,
            data = list(hospital = hospital,community = community),
            nsymp = features,
            n.subset = n.subset,
            prob.wt = prob.wt,
            boot.se = boot.se,
            nboot = nboot,
            printit = printit))
}

#' @importFrom stats sd
#' @export

preprocess <- function(undergrad.results){
  if(is.null(undergrad.results$testset))
    cat("Input must be a list with a 'testset' object\n")
  if(is.null(undergrad.results$trainingset))
    cat("Input must be a list with a 'trainingset' object\n")
  v1 <- apply(as.matrix(undergrad.results$trainingset[, 4:dim(undergrad.results$trainingset)[2]]),2,sd)
  rmc1 <- 3 + which(v1 == 0)
  v2 <- apply(as.matrix(undergrad.results$testset[, 4:dim(undergrad.results$testset)[2]]),2, sd)
  rmc2 <- 3 + which(v2 == 0)
  remove.idx <- c(rmc1,rmc2)
  if(length(remove.idx)>0){
    cat("Removed ",length(remove.idx)," Invariant Columns \n")
    undergrad.preprocessed <- undergrad.results
    undergrad.preprocessed$trainingset <- undergrad.results$trainingset[,-remove.idx]
    undergrad.preprocessed$testset <- undergrad.results$testset[,-remove.idx]
    nnn <- dim(undergrad.preprocessed$trainingset)[1]
    txt <- paste("undergrad.preprocessed$formula <- ",
                 colnames(undergrad.preprocessed$trainingset)[4],"+ ... +",
                 colnames(undergrad.preprocessed$trainingset)[nnn],"~TRUTH",
                 sep = "")
    eval(parse(text=txt))
  }else{
    cat("Removed No Invariant Columns \n")
    undergrad.preprocessed <- undergrad.results
  }
  return(undergrad.preprocessed)
}


#' @importFrom stats as.formula
#' @importFrom utils read.csv write.table
#' @export

undergrad <-
  function(control ="control.txt",stem=T,strip.tags=T,ignore.case=T,table.file="tablefile.txt",threshold=0.01,pyexe=NULL,sep=NULL,printit=TRUE, fullfreq=FALSE, python3=FALSE, alphanumeric.only=TRUE, textincontrol=FALSE, remove.regex=NULL)
{
      os.type <- .Platform$OS.type

      if(is.null(pyexe))
        pyexe <- "python"
          if(python3==FALSE){
            call <- paste("'",system.file("makerfile", package = "ReadMe"),"'",sep="")
      }else{
           call <- paste("'",system.file("makerfile3-0.py", package = "ReadMe"),"'",sep="")
          }

      if(!stem)
        call <- paste(call, "--no-stem")
      if(!strip.tags)
        call <- paste(call, "--tags")
      if(!ignore.case)
        call <- paste(call, "--case-sensitive")
      if(!printit)
        call <- paste(call, "--silent")
      if(alphanumeric.only)
    call <- paste(call, "--alphanumeric-only")
      if(textincontrol)
    call <- paste(call, "--in-control-file")

      if(is.data.frame(control))
      {
        write.table(control, "readmetmpctrl.txt", row.names =FALSE, quote =FALSE)
        control <- "readmetmpctrl.txt"
      }

      if(!is.null(sep))
      {
        call <- paste(call, "--separator", paste('"', sep, '"', sep = ""))
      }

      if(!is.null(remove.regex))
      {
        call <- paste(call, " --remove-regex ", "'", remove.regex, "'", sep="")
      }


      call <- paste(call, "--control-file", paste("'",control,"'",sep=""))
      call <- paste(call, "--table-file", paste("'",table.file,"'",sep=""))
      call <- paste(call, '--threshold', threshold)

      print(paste(pyexe, call))
      sysres <- system(paste(pyexe, call))
      if(sysres == -1)
      {
        if(os.type == "unix")
          stop("Python pyexe must be installed and on system path.")
        # Running Windows, paths sometimes bad try common options

        pyexe <- NULL
        for (i in 10:50)
        {
            pyexe <- paste("/python",i,"/python.exe",sep="")
            if(file.exists(pyexe))
            {
                warning(paste("Python not on path. Using", pyexe))
                break
            }
        }

        if(is.null(pyexe))
        {
          stop("Python pyexe must be installed and on system path.")
        }


        sysres <- system(paste(pyexe,call))
       }
       if(sysres != 0)
       {
         stop("Python module failed. Aborting undergrad.")

       }
      tab <- read.csv(table.file)

  #tab <- remove.nonvariant(tab)

  # Return List
  ret <- list()
  ret$trainingset <- tab[tab$TRAININGSET==1,]
  ret$testset <- tab[tab$TRAININGSET==0,]

  cnames <- colnames(tab)
  ncols <- length(cnames)
  formula <- paste(cnames[4],"+...+",cnames[ncols],"~TRUTH",sep="")
  formula <- as.formula(formula)

  ret$formula <- formula
  ret$features <- 15
  ret$n.subset <- 300
  ret$prob.wt <- 1
  ret$boot.se <- FALSE
  ret$nboot = 300
  ret$printit = printit

  return (ret)
}


inst/makerfile3-0.py
#!/usr/bin/python
import time
import sys
import re
import getopt

"""Algorithm ported to Python 3.0 by Anton Strezhnev"""

"""Porter Stemming Algorithm
This is the Porter stemming algorithm, ported to Python from the
version coded up in ANSI C by the author. It may be be regarded
as canonical, in that it follows the algorithm presented in

Porter, 1980, An algorithm for suffix stripping, Program, Vol. 14,
no. 3, pp 130-137,

only differing from it at the points maked --DEPARTURE-- below.

See also http://www.tartarus.org/~martin/PorterStemmer

The algorithm as described in the paper could be exactly replicated
by adjusting the points of DEPARTURE, but this is barely necessary,
because (a) the points of DEPARTURE are definitely improvements, and
(b) no encoding of the Porter stemmer I have seen is anything like
as exact as this version, even with the points of DEPARTURE!

Vivake Gupta (v@nano.com)

Release 1: January 2001
"""

import sys

class PorterStemmer:

    def __init__(self):
        """The main part of the stemming algorithm starts here.
        b is a buffer holding a word to be stemmed. The letters are in b[k0],
        b[k0+1] ... ending at b[k]. In fact k0 = 0 in this demo program. k is
        readjusted downwards as the stemming progresses. Zero termination is
        not in fact used in the algorithm.

        Note that only lower case sequences are stemmed. Forcing to lower case
        should be done before stem(...) is called.
        """

        self.b = ""  # buffer for word to be stemmed
        self.k = 0
        self.k0 = 0
        self.j = 0   # j is a general offset into the string

    def cons(self, i):
        """cons(i) is TRUE <=> b[i] is a consonant."""
        if self.b[i] == 'a' or self.b[i] == 'e' or self.b[i] == 'i' or self.b[i] == 'o' or self.b[i] == 'u':
            return 0
        if self.b[i] == 'y':
            if i == self.k0:
                return 1
            else:
                return (not self.cons(i - 1))
        return 1

    def m(self):
        """m() measures the number of consonant sequences between k0 and j.
        if c is a consonant sequence and v a vowel sequence, and <..>
        indicates arbitrary presence,

           <c><v>       gives 0
           <c>vc<v>     gives 1
           <c>vcvc<v>   gives 2
           <c>vcvcvc<v> gives 3
           ....
        """
        n = 0
        i = self.k0
        while 1:
            if i > self.j:
                return n
            if not self.cons(i):
                break
            i = i + 1
        i = i + 1
        while 1:
            while 1:
                if i > self.j:
                    return n
                if self.cons(i):
                    break
                i = i + 1
            i = i + 1
            n = n + 1
            while 1:
                if i > self.j:
                    return n
                if not self.cons(i):
                    break
                i = i + 1
            i = i + 1

    def vowelinstem(self):
        """vowelinstem() is TRUE <=> k0,...j contains a vowel"""
        for i in range(self.k0, self.j + 1):
            if not self.cons(i):
                return 1
        return 0

    def doublec(self, j):
        """doublec(j) is TRUE <=> j,(j-1) contain a double consonant."""
        if j < (self.k0 + 1):
            return 0
        if (self.b[j] != self.b[j-1]):
            return 0
        return self.cons(j)

    def cvc(self, i):
        """cvc(i) is TRUE <=> i-2,i-1,i has the form consonant - vowel - consonant
        and also if the second c is not w,x or y. this is used when trying to
        restore an e at the end of a short  e.g.

           cav(e), lov(e), hop(e), crim(e), but
           snow, box, tray.
        """
        if i < (self.k0 + 2) or not self.cons(i) or self.cons(i-1) or not self.cons(i-2):
            return 0
        ch = self.b[i]
        if ch == 'w' or ch == 'x' or ch == 'y':
            return 0
        return 1

    def ends(self, s):
        """ends(s) is TRUE <=> k0,...k ends with the string s."""
        length = len(s)
        if s[length - 1] != self.b[self.k]: # tiny speed-up
            return 0
        if length > (self.k - self.k0 + 1):
            return 0
        if self.b[self.k-length+1:self.k+1] != s:
            return 0
        self.j = self.k - length
        return 1

    def setto(self, s):
        """setto(s) sets (j+1),...k to the characters in the string s, readjusting k."""
        length = len(s)
        self.b = self.b[:self.j+1] + s + self.b[self.j+length+1:]
        self.k = self.j + length

    def r(self, s):
        """r(s) is used further down."""
        if self.m() > 0:
            self.setto(s)

    def step1ab(self):
        """step1ab() gets rid of plurals and -ed or -ing. e.g.

           caresses  ->  caress
           ponies    ->  poni
           ties      ->  ti
           caress    ->  caress
           cats      ->  cat

           feed      ->  feed
           agreed    ->  agree
           disabled  ->  disable

           matting   ->  mat
           mating    ->  mate
           meeting   ->  meet
           milling   ->  mill
           messing   ->  mess

           meetings  ->  meet
        """
        if self.b[self.k] == 's':
            if self.ends("sses"):
                self.k = self.k - 2
            elif self.ends("ies"):
                self.setto("i")
            elif self.b[self.k - 1] != 's':
                self.k = self.k - 1
        if self.ends("eed"):
            if self.m() > 0:
                self.k = self.k - 1
        elif (self.ends("ed") or self.ends("ing")) and self.vowelinstem():
            self.k = self.j
            if self.ends("at"):   self.setto("ate")
            elif self.ends("bl"): self.setto("ble")
            elif self.ends("iz"): self.setto("ize")
            elif self.doublec(self.k):
                self.k = self.k - 1
                ch = self.b[self.k]
                if ch == 'l' or ch == 's' or ch == 'z':
                    self.k = self.k + 1
            elif (self.m() == 1 and self.cvc(self.k)):
                self.setto("e")

    def step1c(self):
        """step1c() turns terminal y to i when there is another vowel in the stem."""
        if (self.ends("y") and self.vowelinstem()):
            self.b = self.b[:self.k] + 'i' + self.b[self.k+1:]

    def step2(self):
        """step2() maps double suffices to single ones.
        so -ization ( = -ize plus -ation) maps to -ize etc. note that the
        string before the suffix must give m() > 0.
        """
        if self.b[self.k - 1] == 'a':
            if self.ends("ational"):   self.r("ate")
            elif self.ends("tional"):  self.r("tion")
        elif self.b[self.k - 1] == 'c':
            if self.ends("enci"):      self.r("ence")
            elif self.ends("anci"):    self.r("ance")
        elif self.b[self.k - 1] == 'e':
            if self.ends("izer"):      self.r("ize")
        elif self.b[self.k - 1] == 'l':
            if self.ends("bli"):       self.r("ble") # --DEPARTURE--
            # To match the published algorithm, replace this phrase with
            #   if self.ends("abli"):      self.r("able")
            elif self.ends("alli"):    self.r("al")
            elif self.ends("entli"):   self.r("ent")
            elif self.ends("eli"):     self.r("e")
            elif self.ends("ousli"):   self.r("ous")
        elif self.b[self.k - 1] == 'o':
            if self.ends("ization"):   self.r("ize")
            elif self.ends("ation"):   self.r("ate")
            elif self.ends("ator"):    self.r("ate")
        elif self.b[self.k - 1] == 's':
            if self.ends("alism"):     self.r("al")
            elif self.ends("iveness"): self.r("ive")
            elif self.ends("fulness"): self.r("ful")
            elif self.ends("ousness"): self.r("ous")
        elif self.b[self.k - 1] == 't':
            if self.ends("aliti"):     self.r("al")
            elif self.ends("iviti"):   self.r("ive")
            elif self.ends("biliti"):  self.r("ble")
        elif self.b[self.k - 1] == 'g': # --DEPARTURE--
            if self.ends("logi"):      self.r("log")
        # To match the published algorithm, delete this phrase

    def step3(self):
        """step3() dels with -ic-, -full, -ness etc. similar strategy to step2."""
        if self.b[self.k] == 'e':
            if self.ends("icate"):     self.r("ic")
            elif self.ends("ative"):   self.r("")
            elif self.ends("alize"):   self.r("al")
        elif self.b[self.k] == 'i':
            if self.ends("iciti"):     self.r("ic")
        elif self.b[self.k] == 'l':
            if self.ends("ical"):      self.r("ic")
            elif self.ends("ful"):     self.r("")
        elif self.b[self.k] == 's':
            if self.ends("ness"):      self.r("")

    def step4(self):
        """step4() takes off -ant, -ence etc., in context <c>vcvc<v>."""
        if self.b[self.k - 1] == 'a':
            if self.ends("al"): pass
            else: return
        elif self.b[self.k - 1] == 'c':
            if self.ends("ance"): pass
            elif self.ends("ence"): pass
            else: return
        elif self.b[self.k - 1] == 'e':
            if self.ends("er"): pass
            else: return
        elif self.b[self.k - 1] == 'i':
            if self.ends("ic"): pass
            else: return
        elif self.b[self.k - 1] == 'l':
            if self.ends("able"): pass
            elif self.ends("ible"): pass
            else: return
        elif self.b[self.k - 1] == 'n':
            if self.ends("ant"): pass
            elif self.ends("ement"): pass
            elif self.ends("ment"): pass
            elif self.ends("ent"): pass
            else: return
        elif self.b[self.k - 1] == 'o':
            if self.ends("ion") and (self.b[self.j] == 's' or self.b[self.j] == 't'): pass
            elif self.ends("ou"): pass
            # takes care of -ous
            else: return
        elif self.b[self.k - 1] == 's':
            if self.ends("ism"): pass
            else: return
        elif self.b[self.k - 1] == 't':
            if self.ends("ate"): pass
            elif self.ends("iti"): pass
            else: return
        elif self.b[self.k - 1] == 'u':
            if self.ends("ous"): pass
            else: return
        elif self.b[self.k - 1] == 'v':
            if self.ends("ive"): pass
            else: return
        elif self.b[self.k - 1] == 'z':
            if self.ends("ize"): pass
            else: return
        else:
            return
        if self.m() > 1:
            self.k = self.j

    def step5(self):
        """step5() removes a final -e if m() > 1, and changes -ll to -l if
        m() > 1.
        """
        self.j = self.k
        if self.b[self.k] == 'e':
            a = self.m()
            if a > 1 or (a == 1 and not self.cvc(self.k-1)):
                self.k = self.k - 1
        if self.b[self.k] == 'l' and self.doublec(self.k) and self.m() > 1:
            self.k = self.k -1

    def stem(self, p, i, j):
        """In stem(p,i,j), p is a char pointer, and the string to be stemmed
        is from p[i] to p[j] inclusive. Typically i is zero and j is the
        offset to the last character of a string, (p[j+1] == '\0'). The
        stemmer adjusts the characters p[i] ... p[j] and returns the new
        end-point of the string, k. Stemming never increases word length, so
        i <= k <= j. To turn the stemmer into a module, declare 'stem' as
        extern, and delete the remainder of this file.
        """
        # copy the parameters into statics
        self.b = p
        self.k = j
        self.k0 = i
        if self.k <= self.k0 + 1:
            return self.b # --DEPARTURE--

        # With this line, strings of length 1 or 2 don't go through the
        # stemming process, although no mention is made of this in the
        # published algorithm. Remove the line to match the published
        # algorithm.

        self.step1ab()
        self.step1c()
        self.step2()
        self.step3()
        self.step4()
        self.step5()
        return self.b[self.k0:self.k+1]


def getControlData(controlfilename,sep):                   
    try:
        controlfile = open(controlfilename, encoding='utf-8')
        controlfile.readline() # skip first line (header)
    except:
        print("Control file missing or empty.")
        sys.exit(1)
    controldata = {'filename' : [], 'truth' : [], 'trainingset' : []}
    while True:
        try: line = controlfile.readline().strip()
        except: break
        if not line:
            break
        try:
            
            (filename, truth, trainingset) = line.split(sep)
        except ValueError:
            print("Control file not properly formatted. Each line must contain three %s-delimited columns. Check separator settings." % (sep or "tab"))
            sys.exit(1)
        controldata['filename'].append(filename)
        controldata['truth'].append(truth)
        controldata['trainingset'].append(trainingset)
		
    return controldata
    
def getText(filename, infile):
    if infile:
        text = filename.rstrip('"').lstrip('"')
    else:
        f = open(filename, encoding='utf-8')
        text = ' '.join(f.readlines())
    if not text: return ' '
    return text

def striptags(posttext):
 linkre = re.compile(r'<.+?>', re.S|re.I)
 posttext = re.sub(r'<br.+?>',r'\n',posttext)
 posttext = re.sub(linkre, r'', posttext)
 return posttext

def getwordset(text,stripTags,stem,ignoreCase,alphanumOnly, removeRegex):
  stemmer = PorterStemmer()
  if ignoreCase:
      text = text.lower()
  if stripTags:
      text = striptags(text.lower())
  if alphanumOnly:
    text = re.sub('\W',' ',text)
  
  if removeRegex != None:
    text = re.sub(removeRegex, ' ', text)
    
  dict = {}
  splittext = text.split()
  if stem:
      splittext = [stemmer.stem(x,0,len(x)-1) for x in splittext]
      
  for word in splittext:
      if word in dict:
          dict[word] += 1
      else: dict[word] = 1

  return dict

def getallwords(controldata, stripTags, stem, ignoreCase, threshold, alphanumOnly, inFile, removeRegex):

 seen = {}
 count = 0.
 for filename in controldata['filename']:
  count += 1
  text = getText(filename, inFile)
  
  if not text: text = ""
  words = getwordset(text,stripTags,stem,ignoreCase, alphanumOnly, removeRegex)
  for word in words.keys():
   if word in seen:
    seen[word] += 1
   else: seen[word] = 0

 return set([x for x in list(seen.keys()) if seen[x] >= count * threshold and seen[x] < count])
  
def outputdata(controldata,stripTags, stem, ignoreCase, alphanumOnly, inFile, removeRegex,threshold, tablefile,fullfreq,printit):

    allwords = getallwords(controldata, stripTags, stem, ignoreCase, threshold, alphanumOnly, inFile, removeRegex)
    filenames = controldata['filename'] 
    truth = controldata['truth']
    trainingset = controldata['trainingset'] 

    rfile = open(tablefile, "w", encoding="utf-8")
    
    rfile.write('"FILENAME","TRUTH","TRAININGSET",')
    rfile.write(",".join(['"WORD:%s"' % x for x in allwords]))
    rfile.write('\n')
    numfiles = len(filenames)
    if printit:
        print("Processing Texts")
        print("Current File", "Number of Files")
    for i in range(numfiles):

        if printit:
            print(i+1,numfiles)

        filename = filenames[i]
        text = getText(filename, inFile)
        filewords = getwordset(text,stripTags,stem,ignoreCase, alphanumOnly, removeRegex)
        rfile.write('"%s",%s,%s' % (filenames[i], truth[i], trainingset[i]))
        
        for word in allwords:
            if word in filewords:
                if fullfreq:
                    rfile.write(',%d' % filewords[word])
                else: rfile.write(',1')
            else:
                rfile.write(',0')
        rfile.write('\n')
        
            
def main():

    # Defaults
    stem = True
    stripTags = True
    ignoreCase = True
    threshold = 0.01
    controlfile = "control.txt"
    tablefile = "tablefile.txt"
    fullfreq = False
    printit = True
    alphanumOnly = False
    removeRegex = None
    inFile = False
    sep = None
    (args,other) = getopt.getopt(sys.argv[1:], "", ['no-stem','tags','case-sensitive','silent','separator=','threshold=', 'control-file=','table-file=','alphanumeric-only','in-control-file','remove-regex='])


    for (a,v) in args:
        if a == '--no-stem': stem = False
        if a == '--tags': stripTags = False
        if a == '--case-sensitive': ignoreCase = False
        if a == '--threshold': threshold = float(v)
        if a == '--control-file': controlfile = v
        if a == '--table-file': tablefile = v
        if a == '--silent': printit = False
        if a == '--separator': sep = v 
        if a == '--frequency-table': fullfreq = True
        if a == '--alphanumeric-only': alphanumOnly = True
        if a == '--in-control-file': inFile = True
        if a == '--remove-regex': removeRegex = v
    controldata = getControlData(controlfile,sep)
    outputdata(controldata,stripTags, stem, ignoreCase, alphanumOnly, inFile, removeRegex, threshold, tablefile,fullfreq,printit)

 

main()




